src/server/handlers/launchpad.js

```diff
@@ -122,107 +143,137 @@ const responseError = (res, error) => {
-        git_repository_url: repositoryUrl,
-        git_path: 'refs/heads/master',
-        auto_build: true,
-        auto_build_archive: `/${DISTRIBUTION}/+archive/primary`,
-        auto_build_pocket: 'Updates',
-        processors: ARCHITECTURES.map(arch => {
-          return `/+processors/${arch}`;
-        }),
-        store_upload: true,
-        store_series: `/+snappy-series/${STORE_SERIES}`,
-        store_name: snapcraftYaml.name
+  const repositoryUrl = req.body.repository_url;
+  const lp_client = getLaunchpad();
+  let self_link;
+  // We need admin permissions in order to be able to install a webhook later.
+  checkAdminPermissions(req)
+    .then(([owner, name, token]) => getSnapcraftYaml(owner, name, token))
+    .then((snapcraftYaml) => {
+      if (!('name' in snapcraftYaml)) {
+        throw new PreparedError(400, RESPONSE_SNAPCRAFT_YAML_NO_NAME);
       }
-    }).then(result => {
-      logger.info(`Authorizing ${result.self_link}`);
-      return lp_client.named_post(result.self_link, 'beginAuthorization')
-        .then(caveatId => {
-          logger.info(`Began authorization of ${result.self_link}`);
-          return res.status(201).send({
-            status: 'success',
-            payload: {
-              code: 'snap-created',
-              message: caveatId
-            }
-          });
-        });
-    }).catch(error => responseError(res, error));
-  });
+      const username = conf.get('LP_API_USERNAME');
+      logger.info(`Creating new snap for ${repositoryUrl}`);
+      return lp_client.named_post('/+snaps', 'new', {
+        parameters: {
+          owner: `/~${username}`,
+          distro_series: `/${DISTRIBUTION}/${DISTRO_SERIES}`,
+          name: `${makeSnapName(repositoryUrl)}-${DISTRO_SERIES}`,
+          git_repository_url: repositoryUrl,
+          git_path: 'refs/heads/master',
+          auto_build: true,
+          auto_build_archive: `/${DISTRIBUTION}/+archive/primary`,
+          auto_build_pocket: 'Updates',
+          processors: ARCHITECTURES.map((arch) => {
+            return `/+processors/${arch}`;
+          }),
+          store_upload: true,
+          store_series: `/+snappy-series/${STORE_SERIES}`,
+          store_name: snapcraftYaml.name
+        }
+      });
+    })
+    .then((result) => {
+      self_link = result.self_link;
+      logger.info(`Authorizing ${self_link}`);
+      return lp_client.named_post(self_link, 'beginAuthorization');
+    })
+    .then((caveatId) => {
+      logger.info(`Began authorization of ${self_link}`);
+      return res.status(201).send({
+        status: 'success',
+        payload: {
+          code: 'snap-created',
+          message: caveatId
+        }
+      });
+    })
+    .catch((error) => responseError(res, error));
 };
 
-const internalFindSnap = async repositoryUrl => {
+const internalFindSnap = async (repositoryUrl) => {
   const cacheId = `url:${repositoryUrl}`;
 
   return new Promise((resolve, reject) => {
```

Comment:
Why returning `new Promise` when `named_get` is already a Promise?

Just realised `named_get` should also be camelCase in theory... but we don't have to update it all at once.